SmartDeviceLink/SDLIAPTransport.m

```diff
@@ -389,334 +404,210 @@ - (void)sdl_establishSessionWithAccessory:(nullable EAAccessory *)accessory {
+/**
+ *  Called when data is received during the data session. The data is passed to the listeners.
+ *
+ *  @return A SDLIAPDataSessionCreateDataReceivedHandler handler
+ */
+- (nullable SDLIAPDataSessionDataReceivedHandler)dataReceivedHandler {
     __weak typeof(self) weakSelf = self;
-    
-    return ^(NSInputStream *istream) {
-        __strong typeof(weakSelf) strongSelf = weakSelf;
-        SDLLogV(@"Control stream received data");
-        
-        // Read in the stream a single byte at a time
-        uint8_t buf[1];
-        NSInteger len = [istream read:buf maxLength:1];
-        if (len <= 0) {
-            return;
-        }
-        
-        // If we have data from the stream
-        // Determine protocol string of the data session, then create that data session
-        NSString *indexedProtocolString = [NSString stringWithFormat:@"%@%@", IndexedProtocolStringPrefix, @(buf[0])];
-        SDLLogD(@"Control Stream will switch to protocol %@", indexedProtocolString);
-        
-        // Destroy the control session
-        dispatch_sync(dispatch_get_main_queue(), ^{
-            [strongSelf.controlSession stop];
-            strongSelf.controlSession.streamDelegate = nil;
-            strongSelf.controlSession = nil;
-        });
-        
-        if (accessory.isConnected) {
-            dispatch_async(dispatch_get_main_queue(), ^{
-                [strongSelf sdl_createIAPDataSessionWithAccessory:accessory forProtocol:indexedProtocolString];
-                [strongSelf.protocolIndexTimer cancel];
-            });
-        }
-
-        [strongSelf sdl_backgroundTaskStart];
+    return ^(NSData * _Nonnull dataIn) {
+        [weakSelf.delegate onDataReceived:dataIn];
+        [weakSelf sdl_backgroundTaskStart];
     };
 }
 
-- (SDLStreamErrorHandler)sdl_controlStreamErroredHandler {
-    __weak typeof(self) weakSelf = self;
-    
-    return ^(NSStream *stream) {
-        __strong typeof(weakSelf) strongSelf = weakSelf;
-        SDLLogE(@"Control stream error");
-        
-        [strongSelf.protocolIndexTimer cancel];
-        [strongSelf.controlSession stop];
-        strongSelf.controlSession.streamDelegate = nil;
-        strongSelf.controlSession = nil;
-        [strongSelf sdl_retryEstablishSession];
-    };
-}
+#pragma mark - SDLIAPSessionDelegate
+
+/**
+ *  Called after both the input and output streams of the session have opened.
+ *
+ *  @param session The current session
+ */
+- (void)onSessionInitializationCompleteForSession:(SDLIAPSession *)session {
+    SDLLogV(@"%@ I/O streams opened for protocol: %@", ([ControlProtocolString isEqualToString:session.protocol] ? @"Control" : @"Data"), session.protocol);
 
+    if ([ControlProtocolString isEqualToString:session.protocol]) {
+        if (!self.dataSession.session) {
+            [self.controlSession startSessionTimer];
+        }
+    }
 
-#pragma mark Data Stream
+    if (![ControlProtocolString isEqualToString:session.protocol]) {
```

Comment:
Wouldn't this just be an `else` to the above `if`?

Reply:
Fixed